apis/core/v1alpha2/featuregate_webhook.go

```diff
@@ -0,0 +1,327 @@
+		return nil
+	}
+
+	c, err := r.getClient()
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	ctx := context.Background()
+	var allErrors field.ErrorList
+
+	allErrors = append(allErrors, r.validateFeatureExists(ctx, c)...)
+	allErrors = append(allErrors, r.validateConflictingFeaturesInFeatureGate(ctx, c)...)
+	allErrors = append(allErrors, r.validateWarrantyVoidOverride(oldObj)...)
+	allErrors = append(allErrors, r.validateFeatureForStabilityPolicyViolation(ctx, c)...)
+
+	if len(allErrors) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("FeatureGate").GroupKind(), r.Name, allErrors)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateDelete() error {
+	featuregatelog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+// validateFeatureForStabilityPolicyViolation validates features for any stability policy violation in a FeatureGate
+// resource
+func (r *FeatureGate) validateFeatureForStabilityPolicyViolation(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	features := &FeatureList{}
+	if err := c.List(ctx, features); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	featuresThatVoidWarranty := computeFeaturesThatVoidSupportWarranty(r.Spec, features)
+	immutableFeatures := computeImmutableFeatures(r.Spec, features)
+
+	if len(featuresThatVoidWarranty) > 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"),
+			r.Spec.Features, fmt.Sprintf("cannot toggle features %v as the the stability level of the "+
+				"features indicate that it should not be activated in production environments. To activate the "+
+				"feature, you must agree to permanently void all support guarantees for this environment by setting "+
+				"featureRef.permanentlyVoidAllSupportGuarantees to true", featuresThatVoidWarranty)))
+	}
+
+	if len(immutableFeatures) > 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"),
+			r.Spec.Features, fmt.Sprintf("cannot toggle immutable features: %v", immutableFeatures)))
+	}
+
+	return allErrors
+}
+
+// computeFeaturesThatVoidSupportWarranty computes and returns features that voids the support warranty in a FeatureGate
+// resource spec
+func computeFeaturesThatVoidSupportWarranty(spec FeatureGateSpec, features *FeatureList) []string {
+	invalidFeatures := sets.String{}
+	for _, featureRef := range spec.Features {
+		stabilityLevel, found := getFeatureStabilityLevel(features, featureRef.Name)
+		if !found {
+			// Feature doesn't exist and is validated in validateFeatureExistence method
+			continue
+		}
+		policy := GetPolicyForStabilityLevel(stabilityLevel)
+		if policy.VoidsWarranty && !featureRef.PermanentlyVoidAllSupportGuarantees &&
+			policy.DefaultActivation != featureRef.Activate {
```

Comment:
question: trying to understand the logic of `policy.VoidsWarranty && !featureRef.PermanentlyVoidAllSupportGuarantees`, is it equivalent to `policy.VoidsWarranty != featureRef.PermanentlyVoidAllSupportGuarantees` ?

And another question is why we need `policy.DefaultActivation != featureRef.Activate` ?

Reply:
ya we could write that as `policy.VoidsWarranty != featureRef.PermanentlyVoidAllSupportGuarantees`, the outcome is going to be the same, but i think is more readable and easy to understand `policy.VoidsWarranty && !featureRef.PermanentlyVoidAllSupportGuarantees`, it basically says if the policy voids warranty and if PermanentlyVoidAllSupportGuarantees is not set to true, its not a valid feature ref in the featuregate spec.

regarding `policy.DefaultActivation != featureRef.Activate`, we want to indicate that a feature is invalid only when toggling the feature, for example, we don't want to say the feature is invalid in the below case 

```
apiVersion: core.tanzu.vmware.com/v1alpha2
kind: FeatureGate
metadata:
  name: tkg-system
spec:
  features:
    - name: big-cache
      activate: false
---
apiVersion: core.tanzu.vmware.com/v1alpha2
kind: Feature
metadata:
  name: big-cache
spec:
  description: "A sample big cache Feature"
  stability: "Experimental"
status:
  activated: false
```

Reply:
could you help add above explanation as inline comment around L169 ? It would help future maintainers to understand the logic better. thanks